### PR TITLE
May 2020 Hackathon RTRIM, LTRIM, RPAD, LPAD Functions

### PR DESCRIPTION
**Pull request by jkosh44**
This PR implements the following built-in string functions:
- ltrim
- rtrim
- lpad
- rpad

I accomplished this by cherry picking the following commits and applying some manual updates:
- https://github.com/apavlo/terrier/pull/17/commits/efb8a9b9f69165a04162304e173b7c2a52eeb747
- https://github.com/apavlo/terrier/pull/17/commits/7f90907f4e7687b4bab4281d9f98a0a0894682d5
- https://github.com/apavlo/terrier/pull/17/commits/cc1e5ce1307f93e9c5050f0777dbdcb8421259a3
- https://github.com/apavlo/terrier/pull/17/commits/141bbca4858e60d03e63b5bc65eccc6513b82aee
- https://github.com/apavlo/terrier/pull/17/commits/6c0b5293f3205db6ea34dba818c2728cd029956f
- https://github.com/apavlo/terrier/pull/17/commits/684453213a971202e0006827d48fe735f70e1843
- https://github.com/apavlo/terrier/pull/17/commits/527151e1b8a27782ba94c6a222602f0f346d8ea5
- https://github.com/apavlo/terrier/pull/17/commits/7d227889ffdf116e10cd9eb335dfb327177096ee

The original PR can be found here: https://github.com/apavlo/terrier/pull/17

Works towards resolving #1001

### CODE REVIEW

File: src/catalog/database_catalog.cpp [hunk 1]
```diff
@@ -2070,6 +2098,40 @@ void DatabaseCatalog::BootstrapProcContexts(const common::ManagedPointer<transac
   SetProcCtxPtr(txn, postgres::TRIM2_PRO_OID, func_context);
   txn->RegisterAbortAction([=]() { delete func_context; });
 
+  func_context = new execution::functions::FunctionContext(
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/catalog/database_catalog.cpp:2101]: I'll do it as part of this PR

File: src/execution/sema/sema_builtin.cpp [hunk 2]
```diff
@@ -2676,6 +2676,88 @@ void Sema::CheckBuiltinStringCall(ast::CallExpr *call, ast::Builtin builtin) {
...
+
+      // checking to see if the first argument is an execution context
+      auto exec_ctx_kind = ast::BuiltinType::ExecutionContext;
+      if (!IsPointerToSpecificBuiltin(call->Arguments()[0]->GetType(), exec_ctx_kind)) {
+        ReportIncorrectCallArg(call, 0, GetBuiltinType(exec_ctx_kind)->PointerTo());
+        return;
+      }
+
+      // checking to see if the second argument is a string
+      auto *resolved_type = Resolve(call->Arguments()[1]);
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/execution/sema/sema_builtin.cpp]: Done

File: src/execution/sema/sema_builtin.cpp [hunk 3]
```diff
@@ -2676,6 +2676,88 @@ void Sema::CheckBuiltinStringCall(ast::CallExpr *call, ast::Builtin builtin) {
...
+        return;
+      }
+      if (!resolved_type->IsSpecificBuiltin(ast::BuiltinType::StringVal)) {
+        ReportIncorrectCallArg(call, 1, ast::StringType::Get(GetContext()));
+        return;
+      }
+
+      if (call->NumArgs() == 3) {
+        // checking to see if the third argument is a string
+        resolved_type = Resolve(call->Arguments()[2]);
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/execution/sema/sema_builtin.cpp]: Done

File: src/execution/sema/sema_builtin.cpp [hunk 4]
```diff
@@ -2676,6 +2676,88 @@ void Sema::CheckBuiltinStringCall(ast::CallExpr *call, ast::Builtin builtin) {
...
+      }
+
+      // checking to see if the third argument is an integer
+      if (call->Arguments()[2]->GetType() != GetBuiltinType(ast::BuiltinType::Integer)) {
+        ReportIncorrectCallArg(call, 2, GetBuiltinType(ast::BuiltinType::Integer));
+        return;
+      }
+
+      // checking to see if the fourth argument is a string
+      resolved_type = Resolve(call->Arguments()[3]);
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/execution/sema/sema_builtin.cpp]: Done

File: src/execution/sema/sema_builtin.cpp [hunk 5]
```diff
@@ -2676,6 +2676,88 @@ void Sema::CheckBuiltinStringCall(ast::CallExpr *call, ast::Builtin builtin) {
...
+
+      // checking to see if the first argument is an execution context
+      auto exec_ctx_kind = ast::BuiltinType::ExecutionContext;
+      if (!IsPointerToSpecificBuiltin(call->Arguments()[0]->GetType(), exec_ctx_kind)) {
+        ReportIncorrectCallArg(call, 0, GetBuiltinType(exec_ctx_kind)->PointerTo());
+        return;
+      }
+
+      // checking to see if the second argument is a string
+      auto *resolved_type = Resolve(call->Arguments()[1]);
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/execution/sema/sema_builtin.cpp]: Done

File: src/catalog/database_catalog.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/catalog/database_catalog.cpp:2111]: Added

File: src/execution/sema/sema_builtin.cpp [hunk 6]
```diff
@@ -2676,6 +2676,78 @@ void Sema::CheckBuiltinStringCall(ast::CallExpr *call, ast::Builtin builtin) {
...
+
+      // checking to see if the first argument is an execution context
+      auto exec_ctx_kind = ast::BuiltinType::ExecutionContext;
+      if (!IsPointerToSpecificBuiltin(call->Arguments()[0]->GetType(), exec_ctx_kind)) {
+        ReportIncorrectCallArg(call, 0, GetBuiltinType(exec_ctx_kind)->PointerTo());
+        return;
+      }
+
+      // checking to see if the second argument is a string
+      auto *resolved_type = call->Arguments()[1]->GetType();
```
Comment (pmenon) [inline src/execution/sema/sema_builtin.cpp:2693]: Remove the `resolved_type` variable and directly access `call->Arguments()`.

File: src/execution/sema/sema_builtin.cpp [hunk 7]
```diff
@@ -2676,6 +2676,78 @@ void Sema::CheckBuiltinStringCall(ast::CallExpr *call, ast::Builtin builtin) {
...
+      // checking to see if the third argument is an integer
+      if (call->Arguments()[2]->GetType() != GetBuiltinType(ast::BuiltinType::Integer)) {
+        ReportIncorrectCallArg(call, 2, GetBuiltinType(ast::BuiltinType::Integer));
+        return;
+      }
+
+      if(call->NumArgs() == 4) {
+        // checking to see if the fourth argument is a string
+        resolved_type = call->Arguments()[3]->GetType();
+        if (call->NumArgs() == 4 && !resolved_type->IsSpecificBuiltin(ast::BuiltinType::StringVal)) {
```
Comment (pmenon) [inline src/execution/sema/sema_builtin.cpp:2708]: Can you remove the duplicate condition and merge the `if`s?

File: src/execution/sema/sema_builtin.cpp [hunk 8]
```diff
@@ -2676,6 +2676,78 @@ void Sema::CheckBuiltinStringCall(ast::CallExpr *call, ast::Builtin builtin) {
...
+
+      // checking to see if the first argument is an execution context
+      auto exec_ctx_kind = ast::BuiltinType::ExecutionContext;
+      if (!IsPointerToSpecificBuiltin(call->Arguments()[0]->GetType(), exec_ctx_kind)) {
+        ReportIncorrectCallArg(call, 0, GetBuiltinType(exec_ctx_kind)->PointerTo());
+        return;
+      }
+
+      // checking to see if the second argument is a string
+      auto *resolved_type = call->Arguments()[1]->GetType();
```
Comment (pmenon) [inline src/execution/sema/sema_builtin.cpp:2732]: Ditto.

**Comment by jkosh44** [comment]
@pmenon I agree on the optional argument comment. I was actually trying to get the bytecode generator to generate the same bytecode and provide a default pad for the 2 arg version, but I couldn't figure out how to get it to work. Where is the binder `pg_catalog.pg_proc`? I'm not exactly sure what that is.

File: script/testing/junit/traces/functions.test [hunk 9]
```diff
@@ -540,5 +540,105 @@ SELECT split_part(str_a_val, 'd', 1) AS result FROM functions1 WHERE is_null = 1
 ----
 1 values hashing to 68b329da9893e34099c7d8ad5cb9c940
 
+query T nosort
+SELECT rpad(str_a_val, 0, str_i_val) AS result FROM functions1 WHERE is_null = 0
```
Comment (pmenon) [inline script/testing/junit/traces/functions.test:544]: Can you add tests that don't include the padding characters.

Comment (jkosh44) [reply script/testing/junit/traces/functions.test:544]: Done

**Pull request merged**